Fixed Clang -O2 optimization problem

r3dunlop opened the pull request:
- This turned out to be caused by a slew std::string = const char* that clang optimized and did not do the proper casting. I think this is a strict aliasing-like problem (I could be wrong, my comp-sci lingo isn't perfect). I have corrected the code to use only const char*. 
- I assume this problem might carry through into epics and other odb read things,. I'll keep my eye out, but other mac users should as well.

r3dunlop commented:
@pcbend, I changed the logic around a bit in the infinite while loop. Can you see any pitfalls that I may have missed?

pcbend commented:
So i might be missing something, i've been staring at a computer for a long
time.   Does this actually search the nodes?  It looks like it checks the
first one and either it works or it breaks the loop...

i can look again later when my eyes are a bit fresher.


On Mon, May 4, 2015 at 8:26 PM, Ryan Dunlop <notifications@github.com>
wrote:

> @pcbend <https://github.com/pcbend>, I changed the logic around a bit in
> the infinite while loop. Can you see any pitfalls that I may have missed?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GRIFFINCollaboration/GRSISort/pull/306#issuecomment-98890619>
> .
>

r3dunlop commented:
You are correct, I think that first break should be a continue.

> On May 4, 2015, at 8:59 PM, pcbend <notifications@github.com> wrote:
> 
> So i might be missing something, i've been staring at a computer for a long
> time. Does this actually search the nodes? It looks like it checks the
> first one and either it works or it breaks the loop...
> 
> i can look again later when my eyes are a bit fresher.
> 
> 
> On Mon, May 4, 2015 at 8:26 PM, Ryan Dunlop <notifications@github.com>
> wrote:
> 
> > @pcbend <https://github.com/pcbend>, I changed the logic around a bit in
> > the infinite while loop. Can you see any pitfalls that I may have missed?
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/GRIFFINCollaboration/GRSISort/pull/306#issuecomment-98890619>
> > .
> >
> —
> Reply to this email directly or view it on GitHub <https://github.com/GRIFFINCollaboration/GRSISort/pull/306#issuecomment-98895887>.
>

pcbend commented:
i think that would work too.

On Mon, May 4, 2015 at 9:00 PM, Ryan Dunlop <notifications@github.com>
wrote:

> You are correct, I think that first break should be a continue.
>
> > On May 4, 2015, at 8:59 PM, pcbend <notifications@github.com> wrote:
> >
> > So i might be missing something, i've been staring at a computer for a
> long
> > time. Does this actually search the nodes? It looks like it checks the
> > first one and either it works or it breaks the loop...
> >
> > i can look again later when my eyes are a bit fresher.
> >
> >
> > On Mon, May 4, 2015 at 8:26 PM, Ryan Dunlop <notifications@github.com>
> > wrote:
> >
> > > @pcbend <https://github.com/pcbend>, I changed the logic around a bit
> in
> > > the infinite while loop. Can you see any pitfalls that I may have
> missed?
> > >
> > > —
> > > Reply to this email directly or view it on GitHub
> > > <
> https://github.com/GRIFFINCollaboration/GRSISort/pull/306#issuecomment-98890619
> >
> > > .
> > >
> > —
> > Reply to this email directly or view it on GitHub <
> https://github.com/GRIFFINCollaboration/GRSISort/pull/306#issuecomment-98895887
> >.
>
> >
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GRIFFINCollaboration/GRSISort/pull/306#issuecomment-98895996>
> .
>

r3dunlop commented:
I've removed that bad break entirely and it looks like it fixed allllll of the problems. Good catch!

pcbend commented:
excellent, nicely done that xml stuff is a pain.

On Mon, May 4, 2015 at 9:04 PM, Ryan Dunlop <notifications@github.com>
wrote:

> I've removed that bad break entirely and it looks like it fixed allllll of
> the problems. Good catch!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GRIFFINCollaboration/GRSISort/pull/306#issuecomment-98896259>
> .
>

r3dunlop commented:
I spoke a little too soon. For some reason our O1 is now dying because node isn’t being set… This comes before the issue I fixed…and O1 used to work. That means there was an error before that the compiler was somehow compiling away…

> On May 4, 2015, at 9:21 PM, pcbend <notifications@github.com> wrote:
> 
> excellent, nicely done that xml stuff is a pain.
> 
> On Mon, May 4, 2015 at 9:04 PM, Ryan Dunlop <notifications@github.com>
> wrote:
> 
> > I've removed that bad break entirely and it looks like it fixed allllll of
> > the problems. Good catch!
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/GRIFFINCollaboration/GRSISort/pull/306#issuecomment-98896259>
> > .
> >
> —
> Reply to this email directly or view it on GitHub <https://github.com/GRIFFINCollaboration/GRSISort/pull/306#issuecomment-98899120>.
>

r3dunlop commented:
This is the function that is killing us. 
```c++
const char* TXMLOdb::GetNodeName(TXMLNode *node) {
   TList *list = node->GetAttributes();
   if(list) {
      std::string buffer = ((TXMLAttr*)(list->At(0)))->GetValue();
      //list->Delete();
      return buffer.c_str();
   } else {
      return "";
   }
}
```
Am I wrong in saying that I don't think this code should always work? Or do strings do something special in memory like smart pointers?

pcbend commented:
No you are right, that shouldn't work.  The string will be destroyed upon
returning making the data in it undefined.

I guess we got luck so far that the compilers we have been using have
handled it ok.
On May 4, 2015 10:09 PM, "Ryan Dunlop" <notifications@github.com> wrote:

> This is the function that is killing us.
>
> const char* TXMLOdb::GetNodeName(TXMLNode *node) {
>    TList *list = node->GetAttributes();
>    if(list) {
>       std::string buffer = ((TXMLAttr*)(list->At(0)))->GetValue();
>       //list->Delete();
>       return buffer.c_str();
>    } else {
>       return "";
>    }
> }
>
> Am I wrong in saying that I don't think this code should always work? Or
> do strings do something special in memory like smart pointers?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/GRIFFINCollaboration/GRSISort/pull/306#issuecomment-98905648>
> .
>

r3dunlop commented:
Ok, I think all of the code I changed would have worked the way it was if this function was ok. I’m going to make this function return a string, we should be good to go after that.

> On May 4, 2015, at 10:18 PM, pcbend <notifications@github.com> wrote:
> 
> No you are right, that shouldn't work. The string will be destroyed upon
> returning making the data in it undefined.
> 
> I guess we got luck so far that the compilers we have been using have
> handled it ok.
> On May 4, 2015 10:09 PM, "Ryan Dunlop" <notifications@github.com> wrote:
> 
> > This is the function that is killing us.
> >
> > const char* TXMLOdb::GetNodeName(TXMLNode *node) {
> > TList *list = node->GetAttributes();
> > if(list) {
> > std::string buffer = ((TXMLAttr*)(list->At(0)))->GetValue();
> > //list->Delete();
> > return buffer.c_str();
> > } else {
> > return "";
> > }
> > }
> >
> > Am I wrong in saying that I don't think this code should always work? Or
> > do strings do something special in memory like smart pointers?
> >
> > —
> > Reply to this email directly or view it on GitHub
> > <https://github.com/GRIFFINCollaboration/GRSISort/pull/306#issuecomment-98905648>
> > .
> >
> —
> Reply to this email directly or view it on GitHub <https://github.com/GRIFFINCollaboration/GRSISort/pull/306#issuecomment-98908718>.
>

r3dunlop commented:
I think it's very easy to get sloppy with const char* because everyone thinks of them as strings and not as pointers (I do anyway). We just have to keep an eye out for them and only return them when they are data members (I think it's easy to get sloppy with mallocs in these situations).
